README.md
# Story Progress

[![pub package](https://img.shields.io/pub/v/story_progress.svg)](https://pub.dartlang.org/packages/story_progress)

A package provides an easy way to show horizontal progress like instagram stories in Flutter project

<p>
    <img src="https://github.com/a914-gowtham/story_progress/blob/master/screenshots/demo1.gif?raw=true?size=200" width="250" height="480"/>
</p>

## How to use

```dart
import 'package:story_progress/story_progress.dart';

```

```dart
  var _formKey = GlobalKey<StoryProgressState>(); 
  var _play = false;

StoryProgress(
       key: _formKey,
       progressCount: 4,
       width: width,
       duration: Duration(seconds: 3),
       onStatusChanged: (value) {
       WidgetsBinding.instance.addPostFrameCallback((_) {
          switch (value) {
             case Status.next:
              //       
             break;
             case Status.previous:
              //
             break;
             case Status.completed:
              ScaffoldMessenger.of(context).showSnackBar(SnackBar(
                          content: Text('Stories completed'),
                          duration: Duration(seconds: 1),
               ));
             break;
             }
           });
       },
)

```

###  Skip and Previous story
Use this function on onPressed event

```dart

_formKey.currentState.skip();
_formKey.currentState.previous();

```

###  Pause and Resume(start) story
Use this function on onPressed event

```dart

  _formKey.currentState.pause();
  _formKey.currentState.resume(); //

```











CHANGELOG.md
## [0.0.5] - 2020-01-01
* reset progress method added

## [0.0.4] - 2020-01-01
* update documentation

## [0.0.3] - 2020-01-01
* update metadata

## [0.0.2] - 2020-01-01
* update metadata

## [0.0.1] - 2020-01-01
* Initial release.


test/story_progress_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:story_progress/story_progress.dart';

void main() {
  test('adds one to input values', () {
    /*final calculator = Calculator();
    expect(calculator.addOne(2), 3);
    expect(calculator.addOne(-7), -6);
    expect(calculator.addOne(0), 1);
    expect(() => calculator.addOne(null), throwsNoSuchMethodError);*/
  });
}


example/lib/home.dart
import 'package:flutter/material.dart';
import 'package:flutter/rendering.dart';
import 'package:story_progress/story_progress.dart';

class Home extends StatefulWidget {
  Home({Key key}) : super(key: key);

  @override
  _HomeState createState() => _HomeState();
}

class _HomeState extends State<Home> {
  var _play = false;

  var _formKey = GlobalKey<StoryProgressState>();

  var imageList = [
    'assets/images3.jpeg',
    'assets/cyberpunk1.jpeg',
    'assets/cyberpunk2.jpg',
    'assets/images4.jpeg'
  ];
  var _currentIndex = 0;

  @override
  void initState() {
    super.initState();

  }

  @override
  Widget build(BuildContext context) {
    double width = MediaQuery.of(context).size.width;
    return Scaffold(
        body: SafeArea(
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: [
              StoryProgress(
                key: _formKey,
                play: _play,
                progressCount: 4,
                width: width, 
                // duration: Duration(seconds: 3),
                onStatusChanged: (value) {
                  print("onStatusChanged $value");
                  WidgetsBinding.instance.addPostFrameCallback((_) {
                      switch (value) {
                        case Status.next:
                          setState(() {
                            _currentIndex++;
                          });
                          break;
                        case Status.previous:
                          setState(() {
                            _currentIndex--;
                          });
                          break;
                        case Status.completed:
                          ScaffoldMessenger.of(context).showSnackBar(SnackBar(
                            content: Text('Stories completed'),
                            duration: Duration(seconds: 1),
                          ));
                          break;
                      }
                  });
                },
              ),
              Spacer(),
              Image.asset(
                imageList[_currentIndex],
                width: 300,
                height: 200,
              ),
              IconButton(icon: Icon(Icons.skip_next), onPressed: (){
                _formKey.currentState.skip();
              }),
              IconButton(icon: Icon(Icons.skip_previous), onPressed: (){
                _formKey.currentState.previous();
              }),
              Spacer(),
            ],
          ),
        ),
        floatingActionButton: FloatingActionButton(
          child: Icon(_play ? Icons.pause : Icons.play_arrow),
          backgroundColor: Colors.deepPurpleAccent,
          onPressed: () {
            setState(() {
              _play = !_play;
            });
          },
        ));
  }
}



lib/story_progress.dart
library story_progress;

import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';

///[Status] enum used for tracking story changes
enum Status { next, previous, completed }

class Controller {
  void skip() {}

  void previous() {}
}

/// This is the stateful widget that the main application instantiates.
class StoryProgress extends StatefulWidget {
  StoryProgress(
      {Key key,
      this.progressCount = 1,
      this.duration = const Duration(seconds: 8),
      this.width = 200.0,
      @required this.onStatusChanged,
      this.color = Colors.black})
      : super(key: key);

  final int progressCount;

  final Duration duration;

  final double width;

  final Function onStatusChanged;

  final Color color;

  @override
  StoryProgressState createState() => StoryProgressState();

// static StoryProgressState of(BuildContext context) => context.findAncestorStateOfType<StoryProgressState>();
}

/// AnimationControllers can be created with `vsync: this` because of TickerProviderStateMixin.
class StoryProgressState extends State<StoryProgress>
    with TickerProviderStateMixin {
  AnimationController _controller;

  ProgressContainer spinningContainer;

  int _progressCount;

  @override
  void initState() {
    super.initState();

    _progressCount = widget.progressCount;
    _controller = AnimationController(
      duration: widget.duration,
      vsync: this,
    );

    spinningContainer = ProgressContainer(
        progressCount: _progressCount,
        width: widget.width,
        color: widget.color,
        statusChanged: (status) {
          widget.onStatusChanged(status);
        },
        controller: _controller);
  }

  @override
  void dispose() {
    _controller.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return spinningContainer;
  }

  void skip() {
    spinningContainer.forward();
  }

  void previous() {
    spinningContainer.previous();
  }

  void resetCurrentProgress() {
    spinningContainer.resetCurrentProgress();
  }

  void resetAllProgress() {
    spinningContainer.resetAllProgress();
  }

  void pause() {
    if (_controller.isAnimating) _controller.stop();
  }

  void resume() {
    _controller.forward();
  }

  bool isPlaying() => _controller.isAnimating;
}

//ProgressContainer contains all the progress widgets.
///It is using [AnimatedWidget] to animate progress with timer
///For read more about animatedWidget refer this [https://api.flutter.dev/flutter/widgets/AnimatedWidget-class.html]
// ignore: must_be_immutable
class ProgressContainer extends AnimatedWidget implements Controller {
  final int progressCount;
  final double width;
  final AnimationController controller;
  int currentIndex = 0;
  final Function statusChanged;
  bool isCompleted = false;
  final Color color;

  ProgressContainer(
      {Key key,
      this.statusChanged,
      this.progressCount,
      this.color,
      this.width,
      this.controller})
      : super(key: key, listenable: controller);

  Animation<double> get _progress => listenable;

  @override
  Widget build(BuildContext context) {
    double _width = width - 6; //front and back margin
    if (progressCount > 1) {
      _width = _width - (2 * progressCount); //margin b/w two progress
    }
    _width = _width / progressCount;
    return Row(
      mainAxisAlignment: MainAxisAlignment.center,
      children: [
        for (var i = 0; i < progressCount; i++) _buildProgress(_width, i),
      ],
    );
  }

  ///This function navigates to previous story.
  ///Only if [currentIndex] is greater than zero
  @override
  void previous() {
    if (currentIndex > 0) {
      currentIndex--;
      controller.reset();
      controller.forward();
      statusChanged(Status.previous);
      isCompleted = false;
    }
  }

  ///This function navigates to next story.
  ///If current story index is equals to [progressCount]-1 that means it is in last progress,
  ///So, [statusChanged] to Status.completed
  void forward() {
    if (currentIndex == progressCount - 1) {
      print('finished'); //completed
      if (!isCompleted) {
        statusChanged(Status.completed);
        isCompleted = true;
      }
    } else {
      currentIndex++;
      controller.reset();
      controller.forward();
      statusChanged(Status.next);
    }
  }

  void resetCurrentProgress() {
    controller.reset();
  }

  void resetAllProgress() {
    controller.reset();
    currentIndex = 0;
  }

  @override
  void skip() {
    forward();
  }

  ///This function return single progress view
  ///[width] is divided by [progressCount]
  Row _buildProgress(double width, int index) {
    if (_progress.value > 0.99 && controller.isAnimating) {
      if (progressCount == 1 || currentIndex != index) forward();
    }

    return Row(
      children: [
        index == 0
            ? SizedBox(
                width: 3,
              )
            : Container(),
        Stack(
          children: [
            Align(
              alignment: Alignment.centerLeft,
              child: Container(
                decoration: new BoxDecoration(
                  color: color.withOpacity(0.3),
                  shape: BoxShape.rectangle,
                  borderRadius: BorderRadius.all(Radius.circular(8.0)),
                ),
                width: width,
                height: 4,
              ),
            ),
            Align(
              alignment: Alignment.centerLeft,
              child: Container(
                decoration: new BoxDecoration(
                  color: color,
                  shape: BoxShape.rectangle,
                  borderRadius: BorderRadius.all(Radius.circular(8.0)),
                ),
                width: currentIndex == index
                    ? width * _progress.value
                    : (index < currentIndex ? width : 0),
                height: 4,
              ),
            ),
          ],
        ),
        progressCount > 1 && index != progressCount - 1
            ? SizedBox(
                width: 2,
              )
            : SizedBox(
                width: 3,
              ),
      ],
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:flutter/rendering.dart';
import 'package:story_progress/story_progress.dart';

void main() => runApp(MyApp());

/// This is the main application widget.
class MyApp extends StatelessWidget {
  static const String _title = 'Flutter Code Sample';

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: _title,
      home: Home(),
      debugShowCheckedModeBanner: false,
    );
  }
}

class Home extends StatefulWidget {
  Home({Key key}) : super(key: key);

  @override
  _HomeState createState() => _HomeState();
}

class _HomeState extends State<Home> {
  var isPlaying = false;

  var _formKey = GlobalKey<StoryProgressState>();

  var imageList = [
    'assets/images3.jpeg',
    'assets/cyberpunk1.jpeg',
    'assets/cyberpunk2.jpg',
    'assets/images4.jpeg'
  ];
  var _currentIndex = 0;

  @override
  void initState() {
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    double width = MediaQuery.of(context).size.width;
    return Scaffold(
        body: SafeArea(
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: [
              StoryProgress(
                key: _formKey,
                progressCount: 4,
                width: width,
                duration: Duration(seconds: 4),
                onStatusChanged: (value) {
                  print("onStatusChanged $value");
                  WidgetsBinding.instance.addPostFrameCallback((_) {
                    switch (value) {
                      case Status.next:
                        setState(() {
                          _currentIndex++;
                        });
                        break;
                      case Status.previous:
                        setState(() {
                          _currentIndex--;
                        });
                        break;
                      case Status.completed:
                        ScaffoldMessenger.of(context).showSnackBar(SnackBar(
                          content: Text('Stories completed'),
                          duration: Duration(seconds: 2),
                        ));
                        break;
                    }
                  });
                },
              ),
              Spacer(),
              Image.asset(
                imageList[_currentIndex],
                width: 300,
                height: 200,
              ),
              IconButton(
                  icon: Icon(Icons.skip_next),
                  onPressed: () {
                    _formKey.currentState.skip();
                    setState(() {
                      _currentIndex = 0;
                      isPlaying = false;
                    });
                  }),
              IconButton(
                  icon: Icon(Icons.skip_previous),
                  onPressed: () {
                    _formKey.currentState.previous();
                    setState(() {
                      isPlaying = false;
                    });
                  }),
              Spacer(),
            ],
          ),
        ),
        floatingActionButton: FloatingActionButton(
          child: Icon(isPlaying ? Icons.pause : Icons.play_arrow),
          backgroundColor: Colors.deepPurpleAccent,
          onPressed: () {
            if (_formKey.currentState.isPlaying()) {
              _formKey.currentState.pause();
              isPlaying = false;
            } else {
              isPlaying = true;
              _formKey.currentState.resume();
            }
            setState(() {});
          },
        ));
  }
}
